src/Microsoft.Azure.WebJobs.Host/Scale/DefaultHostProcessMonitor.cs

```diff
@@ -0,0 +1,257 @@
+        }
+
+        private HostHealthState GetMemoryStatus(ProcessStats hostProcessStats, IEnumerable<ProcessStats> childProcessStats, ILogger logger = null)
+        {
+            HostHealthState status = HostHealthState.Unknown;
+
+            // if memory throttling is disabled return immediately
+            if (!_options.Value.MemoryThrottleEnabled)
+            {
+                return status;
+            }
+
+            // first compute Memory usage for any registered child processes
+            double currentChildMemoryUsageTotal = 0;
+            foreach (var currentChildStats in childProcessStats.Where(p => p.MemoryUsageHistory.Count() >= MinSampleCount))
+            {
+                // take the last N samples
+                int currChildProcessMemoryStatsCount = currentChildStats.MemoryUsageHistory.Count();
+                var currChildMemoryStats = currentChildStats.MemoryUsageHistory.Skip(currChildProcessMemoryStatsCount - MinSampleCount).Take(MinSampleCount);
+                var currChildMemoryStatsAverage = currChildMemoryStats.Average();
+
+                currentChildMemoryUsageTotal += currChildMemoryStats.Last();
+
+                string formattedLoadHistory = string.Join(",", currChildMemoryStats);
+                logger?.HostProcessMemoryUsage(currentChildStats.ProcessId, formattedLoadHistory, currChildMemoryStatsAverage, currChildMemoryStats.Max());
+            }
+
+            // calculate the aggregate usage across host + child processes
+            int hostProcessMemoryStatsCount = hostProcessStats.MemoryUsageHistory.Count();
+            if (hostProcessMemoryStatsCount >= MinSampleCount)
+            {
+                var lastSamples = hostProcessStats.MemoryUsageHistory.Skip(hostProcessMemoryStatsCount - MinSampleCount).Take(MinSampleCount);
+
+                string formattedUsageHistory = string.Join(",", lastSamples);
+                var hostAverageMemoryUsage = lastSamples.Average();
+                logger?.HostProcessMemoryUsage(hostProcessStats.ProcessId, formattedUsageHistory, Math.Round(hostAverageMemoryUsage), lastSamples.Max());
+
+                // For memory limit, unlike cpu we don't want to compare the average against the limit,
+                // we want to use the last/current. Memory needs to be a harder limit that we want to avoid hitting
+                // otherwise host will see OOM exceptions.
+                double currentMemoryUsage = currentChildMemoryUsageTotal + lastSamples.Last();
+
+                // compute the current total memory usage for host + children
+                var percentageOfMax = (int)(100 * (currentMemoryUsage / _maxMemoryThresholdBytes));
+                logger?.HostAggregateMemoryUsage(currentMemoryUsage, percentageOfMax);
+
+                // if the average is above our threshold, return true (we're overloaded)
+                if (currentMemoryUsage >= _maxMemoryThresholdBytes)
+                {
+                    GC.Collect();
```

Comment:
I added it because in some of my testing, when I configured an arbitrarily low memory limit, once reached the host would stay in that state, likely because usage wasn't actually high enough to trigger collection from the GC's perspective. Given that I've made the memory limit configurable (via ConcurrencyOptions.TotalAvailableMemoryBytes), once the limit is reached, I need a way for the host to reclaim unused memory.

Perhaps if I added a throttle on collections, so that it's done only so often when in this state?

Reply:
I see. Does this mean that the memory usage throttle is set too low? Essentially you've built your own GC logic if we GC every time this throttle is hit, and the GC itself never kicks in, and thats the part I'm uncomfortable with. 

A related question - does the .NET team throttle the threadpool based on available memory? If not, why not? What makes us think that memory based throttling would work for us, but not for the threadpool? 

You also mentioned that the memory throttle rarely kicks in, even in high memory scenarios (other throttles kicked in first). This makes me wonder whether we should keep this throttle.

Reply:
What I've done for now is removed the memory throttle and config from the public interface, and logged https://github.com/Azure/azure-webjobs-sdk/issues/2733. I've left the code/tests in place so while customers can't use it and it's fully disabled, I still retain the integration work I've done while we're deciding. The relevant points in the code reference this new issue. E2E tests use the internal settings to enable and test.

My E2E test "DynamicConcurrencyEnabled_HighMemory_MemoryThrottleEnabled_Throttles" does show the memory throttle kicks in, often in addition to other throttles. While in an in-proc scenario, other throttles might kick in (potentially first), in a OOP case where the memory is being consumed in a child process, things may behave differently. We need more testing there. Definitely the GC call would only help in the in-proc case.

The .NET thread pool algorithm doesn't take memory into account - I'm not sure of the history there.